include/nntile/kernel/cpu.hh

```diff
@@ -0,0 +1,160 @@
+/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
+ *                              (Skoltech), Russia. All rights reserved.
+ *                 2023-present Artificial Intelligence Research Institute
+ *                              (AIRI), Russia. All rights reserved.
+ *
+ * NNTile is software framework for fast training of big neural networks on
+ * distributed-memory heterogeneous systems based on StarPU runtime system.
+ *
+ * @file include/nntile/kernel/cpu.hh
+ * CPU-related compute types
+ *
+ * @version 1.0.0
+ * */
+
+#pragma once
+
+#include <cstdint>
+
+namespace nntile::kernel
+{
+
+//! Templated class to derive compute type for each of basic NNTile types
+template<typename T>
+struct CPUComputeType
+{
+};
+
+//! Compute type for nntile::int64_t type
+template<>
+struct CPUComputeType<int64_t>
+{
+    // nntile::int64_t -> std::int64_t from <cstdint>
+    using value = std::int64_t;
+};
+
+//! Compute type for nntile::int32_t type
+template<>
+struct CPUComputeType<int32_t>
+{
+    // nntile::int32_t -> std::int32_t from <cstdint>
+    using value = std::int32_t;
+};
+
+//! Compute type for nntile::int16_t type
+template<>
+struct CPUComputeType<int16_t>
+{
+    // nntile::int16_t -> std::int16_t from <cstdint>
+    using value = std::int16_t;
+};
+
+//! Compute type for nntile::int8_t type
+template<>
+struct CPUComputeType<int8_t>
+{
+    // nntile::int8_t -> std::int8_t from <cstdint>
+    using value = std::int8_t;
+};
+
+//! Compute type for nntile::bool_t type
+template<>
+struct CPUComputeType<bool_t>
+{
+    // nntile::bool_t -> bool from C++ base
+    using value = bool;
+};
+
+//! Compute type for nntile::fp64_t type
+template<>
+struct CPUComputeType<fp64_t>
+{
+    // nntile::fp64_t -> double from C++ base
+    using value = double;
+};
+
+//! Compute type for nntile::fp32_t type
+template<>
+struct CPUComputeType<fp32_t>
+{
+    // nntile::fp32_t -> double from C++ base
+    using value = float;
+};
+
+//! Compute type for nntile::fp32_fast_tf32_t type
+template<>
+struct CPUComputeType<fp32_fast_tf32_t>
+{
+    // No member `value` here is for a reason: this type shall be manually
+    // converted into computing types, as memory-bound operations shall be done
```

Comment:
Is this conversion expected inside every kernel function?

Reply:
Yes. Every kernel function receives inputs in `nntile::Index` and `nntile::scal_t` for parameters and one of `nntile::int64_t`, `nntile::fp64_t`,... types for tensors. Then, it is up kernel to parse these tensor types into underlying types for each device. `CUDAComputeType<T>` converts our wrapped type into a corresponding CUDA type (e.g., `nntile::bf16_t` into `__nv_bfloat16`). Only once it comes to kernels these types shall be casted into proper ones.